gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```diff
@@ -0,0 +1,501 @@
+package gobblin.tunnel;
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+import com.google.common.collect.Sets;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import org.slf4j.LoggerFactory;
+
+import static java.nio.channels.SelectionKey.OP_READ;
+
+
+/**
+ * @author navteniev@linkedin.com
+ *
+ * Implements a tunnel through a proxy to resource on the internet
+ */
+public class Tunnel {
+
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+  public static final int PROXY_CONNECT_TIMEOUT_MS = 5000;
+
+  private final String _remoteHost;
+  private final int _remotePort;
+  private final String _proxyHost;
+  private final int _proxyPort;
+  private ServerSocketChannel _server;
+  private Thread _thread;
+  private static final ByteBuffer OK_REPLY = ByteBuffer.wrap("HTTP/1.1 200".getBytes());
+  private static final Set<ByteBuffer> OK_REPLIES =
+      Sets.newHashSet(OK_REPLY, ByteBuffer.wrap("HTTP/1.0 200".getBytes()));
+
+  private Tunnel(String remoteHost, int remotePort, String proxyHost, int proxyPort) {
+    _remoteHost = remoteHost;
+    _remotePort = remotePort;
+    _proxyHost = proxyHost;
+    _proxyPort = proxyPort;
+  }
+
+  private Tunnel open() {
+    try {
+      _server = ServerSocketChannel.open().bind(null);
+      startTunnelThread();
+
+      return this;
+    } catch (IOException ioe) {
+      LOG.error("Failed to open the tunnel", ioe);
+    }
+
+    return null; 
+  }
+
+  public int getPort() {
+    SocketAddress localAddress = null;
+    try {
+      if (_server != null && _server.isOpen()) {
+        localAddress = _server.getLocalAddress();
+      }
+      if (localAddress instanceof InetSocketAddress) {
+        return ((InetSocketAddress) localAddress).getPort();
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get tunnel port", e);
+    }
+
+    return -1;
```

Comment:
What does -1 mean? If you are using this as a sentinel can you create a named static constant?